### PR TITLE
[MLIR] Enzyme integration with backpropOp

### LINKED ISSUE #122: [MLIR] Re-enable buffer deallocation
Take into account aliasing when passing memrefs as parameters to other functions and returning them.

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the change**

- Lowering update to unpack the memref
- We apply o2 before Enzyme
- Argmap is now destination-passing style
- The wrapper for Enzyme is removed

- Integration in the Hybrid gradient is added
- Test are updated

- Buffer deallocation is readded, which relied on the removal of the resizable array in this branch.

[sc-36401]
[sc-41703]

**Related isses:** Fixes #122

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-41511]

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (dime10) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:277]: @josh146 does this copyright notice look good?

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:340]: Is this only for the first axis? Are we guaranteeing that memref is only of rank 1?

Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:340]: Actually, it is for the last axis, the axis order start at the last one, don't they?

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (josh146) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:277]: Looks good to me!

File: mlir/lib/Gradient/Utils/GradientShape.cpp
Comment (dime10) [reply mlir/lib/Gradient/Utils/GradientShape.cpp:124]: Did we not code it up to support both tensors and scalars? Current arch supports both

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (pengmai) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:340]: This is for a general MemRef. Consider a `memref<MxNxK>`: the sizes will be `[M, N, K]` and the strides will be `[N*K, K, 1]`. The overall number of elements, `M * N * K`, can be obtained by multiplying the `sizes[0] * strides[0]`.

We can update the comment here to add more detail if it's confusing.

Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:340]: Yes, a comment would be nice. Thanks @pengmai!

File: mlir/include/Gradient/IR/GradientOps.td
Comment (dime10) [reply mlir/include/Gradient/IR/GradientOps.td:97]: Fixed this in #224

File: mlir/test/Gradient/AdjointTest.mlir
Comment (dime10) [reply mlir/test/Gradient/AdjointTest.mlir:1]: After some discussion @pengmai is going to re-add that support before merging this PR :)

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (dime10) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:286]: Minor: would be nice if the name of the function indicated that it is modifying one of the args since it doesn't return anything, like `unpackMemRefAndAppend`.

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (dime10) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:297]: Ah this is really neat, didn't know `MemRefDescriptor` was in the library!

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (dime10) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:358]:
```suggestion
    /// This registers custom allocation and deallocation functions with Enzyme. It creates a
```

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (dime10) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:406]: This could use a docstring. Something like `defineGlobalString` or similar for the function name would also be clearer I think.

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (dime10) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:421]: A docstring would also be useful here, in particular to distinguish it from `insertEnzymeAllocationLike`.

**Review by dime10 — COMMENTED** [review]
Amazing work here, my comments are mainly regarding adding some documentation to the code. I also made some minor tweaks to the backprop op in #224.

File: mlir/lib/Gradient/Utils/GradientShape.cpp
Comment (dime10) [reply mlir/lib/Gradient/Utils/GradientShape.cpp:124]: After some discussion @pengmai is going to re-add that support before merging this PR :)

File: doc/changelog.md
Comment (rmoyard) [reply doc/changelog.md]: Updated with an example

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:161]: Yes correct!

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:187]: I had to add it to have "free\0" instead of free.

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:223]: Updated to use adaptor

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:267]: I think they serve different purposes. The first one register an Enzyme global. The second one creates takes a key and a value, to register the name of a function. We could maybe make the value optional in the second one and have only function. I will check that.

Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:267]: Updated !

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
Comment (rmoyard) [reply mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:270]: Done

File: frontend/catalyst/compiler.py
Comment (rmoyard) [reply frontend/catalyst/compiler.py:202]: @erick-xanadu I let you add an entry, you will describe it better than me

File: mlir/test/Gradient/AdjointTest.mlir
Comment (dime10) [inline mlir/test/Gradient/AdjointTest.mlir:1]: In the final form this test seems to have been rewritten to avoid `f32` and `f128` types. Does Enzyme only support `f64`?

File: mlir/test/Gradient/BufferizationTest.mlir
Comment (dime10) [inline mlir/test/Gradient/BufferizationTest.mlir:50]: This test and the next seem to avoid the dynamic dimension, was there a specific reason for that or did it just come about from removing and re-adding the tests?

File: mlir/test/Gradient/ClassicalJacobianTest.mlir
Comment (dime10) [inline mlir/test/Gradient/ClassicalJacobianTest.mlir:66]: In this file as well I'm just curious whether the changes (wrapping the scalars in tensors) were incidental or for a specific purpose (e.g. to diversify the tests)?

File: mlir/test/Gradient/ConversionBackprop.mlir
Comment (dime10) [inline mlir/test/Gradient/ConversionBackprop.mlir:1]: This one removed one tests case right?

File: mlir/test/Gradient/ParameterShiftTest.mlir
Comment (dime10) [inline mlir/test/Gradient/ParameterShiftTest.mlir:1]: This one also added wrapping to scalars and normalization to f64.

**Review by dime10 — APPROVED** [review]
Thank you for this amazing work 💯 🥳 

I was just wondering what the reason behind some of the changes in the tests were, otherwise this PR is good to go once @erick-xanadu adds his changelog entry :)

File: mlir/test/Gradient/AdjointTest.mlir
Comment (rmoyard) [reply mlir/test/Gradient/AdjointTest.mlir:1]: No but there some tests where the initial type was not matching the resulting type.

File: mlir/test/Gradient/ClassicalJacobianTest.mlir
Comment (rmoyard) [reply mlir/test/Gradient/ClassicalJacobianTest.mlir:66]: I think this was my doing because there was some scalar instead of tensors and before Jacob's PR we where not supporting it

File: mlir/test/Gradient/BufferizationTest.mlir
Comment (rmoyard) [reply mlir/test/Gradient/BufferizationTest.mlir:50]: Updated back

File: mlir/test/Gradient/ConversionBackprop.mlir
Comment (rmoyard) [reply mlir/test/Gradient/ConversionBackprop.mlir:1]: Yes correct, I will try to readd it. @pengmai You remember why it was removed?

File: mlir/test/Gradient/ParameterShiftTest.mlir
Comment (rmoyard) [reply mlir/test/Gradient/ParameterShiftTest.mlir:1]: Same as the answer before the wrapping scalars was because of the non-support and the normalization was because you cannot start with f32 and finish with f64 in Enzyme

File: mlir/test/Gradient/ConversionBackprop.mlir
Comment (rmoyard) [reply mlir/test/Gradient/ConversionBackprop.mlir:1]: I will add one with some float and memref mixed

Comment (pengmai) [reply mlir/test/Gradient/ConversionBackprop.mlir:1]: I don't remember removing it, I think it must've been in the mess of disabling/re-adding tests. It also looks like it needs to be updated.

Comment (rmoyard) [reply mlir/test/Gradient/ConversionBackprop.mlir:1]: Ok no worries I will add one :+1:

Comment (rmoyard) [reply mlir/test/Gradient/ConversionBackprop.mlir:1]: Added!

File: mlir/test/Gradient/AdjointTest.mlir
Comment (dime10) [reply mlir/test/Gradient/AdjointTest.mlir:1]: I think this was on purpose to make as few assumptions as possible on the gradient architecture. After all there is no reason that the relationship from input to outputs could not involve casting, or even be independent (admittedly the gradient would be zero in that case).
Not sure if this would work with Enzyme though since the gradients are accumulated in the argument shadows, whereas normally the result type is the one that would determine the gradient type (think finite-diff for reference).

Comment (pengmai) [reply mlir/test/Gradient/AdjointTest.mlir:1]: So a common assumption of reverse-mode autodiff is that the derivative has the same type as the corresponding primal value.

Consider a function that effectively implements an identity: `f(x: f64) -> f32 { y = truncate(x); return y }`. We'll ignore numerical issues from the truncation and assume that it mathematically correctly implements an identity.

Reverse-mode autodiff ensures that the types of gradients and arguments match, meaning it roughly implements `df(x: f64, dy: f32) -> f64 { return extend(dy) }`. Explicitly, `x` and `dx` are both `f64` while `y` and `dy` are both `f32`.

We can see this in PyTorch:
```python
x = torch.tensor(3.14, dtype=torch.float64, requires_grad=True)
y = x.type(torch.FloatTensor)
y.backward()
print(x.grad) # tensor(1., dtype=torch.float64)
```

Finite diff is notably different because the derivative is expressed as a perturbation in the output, meaning a finite diff result would be of type `f32` in this example. I'm not actually sure which one is "better" in a mathematical purity sense. I suppose in math, everything is a real number and so using floating point numbers at all is already a departure from that.

With all of this said, it's possible we could've had a less drastic normalization of the tests because this constraint is more permissive than "everything has to be f64". I'd guess that it wouldn't be too difficult but it would take a bit of effort to make sure we read the types from the right places.

Comment (dime10) [reply mlir/test/Gradient/AdjointTest.mlir:1]: Thanks for the answer @pengmai, that makes sense!

> I'm not actually sure which one is "better" in a mathematical purity sense. I suppose in math, everything is a real number and so using floating point numbers at all is already a departure from that.

Intuitively I would argue that the output type is the relevant one since mathematically the derivative is defined as the limit over the finite difference, but you're probably right and it doesn't really translate all that well to computing 😅

> With all of this said, it's possible we could've had a less drastic normalization of the tests because this constraint is more permissive than "everything has to be f64". I'd guess that it wouldn't be too difficult but it would take a bit of effort to make sure we read the types from the right places.

The only reason I mentioned it is because when we wrote the finite pass we wanted to make sure it was as robust as possible with few assumptions (that otherwise might pop-up as bugs in the future), and so made sure to grab the types and shapes from the right places.

But it's not a crucial feature I suppose, and it does complicate things that not all methods lead to the same type analysis 🤔 In that case it would be nice to assert any such assumptions though if we make them.

**Pull request merged**
